mParticle-Apple-SDK/Next Gen/MPEventLogging.h

```diff
@@ -0,0 +1,50 @@
+//
+//  MPEventLogging.h
+//  mParticle-Apple-SDK
+//
+//  Created by Ben Baron on 6/29/23.
+//
+
+#import <Foundation/Foundation.h>
+#import "MPComponent.h"
+
+@class MPEvent;
+@class MPBaseEvent;
+@class MPCommerceEvent;
+@class MPNetworkPerformance;
+
+typedef NS_ENUM(NSUInteger, MPExecStatus) {
+    MPExecStatusSuccess = 0,
+    MPExecStatusFail,
+    MPExecStatusMissingParam,
+    MPExecStatusDisabledRemotely,
+    MPExecStatusEnabledRemotely,
+    MPExecStatusOptOut,
+    MPExecStatusDataBeingFetched,
+    MPExecStatusInvalidDataType,
+    MPExecStatusDataBeingUploaded,
+    MPExecStatusServerBusy,
+    MPExecStatusItemNotFound,
+    MPExecStatusDisabledInSettings,
+    MPExecStatusNoConnectivity
+};
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface MPEventLogging : NSObject <MPComponent>
+
+@property (nonatomic, strong, nullable) NSMutableSet<MPEvent *> *eventSet;
+
+- (void)leaveBreadcrumb:(nonnull MPEvent *)event completionHandler:(void (^ _Nonnull)(MPEvent * _Nonnull event, MPExecStatus execStatus))completionHandler;
+- (void)logError:(nullable NSString *)message exception:(nullable NSException *)exception topmostContext:(nullable id)topmostContext eventInfo:(nullable NSDictionary *)eventInfo completionHandler:(void (^ _Nonnull)(NSString * _Nullable message, MPExecStatus execStatus))completionHandler;
+- (void)logCrash:(nullable NSString *)message stackTrace:(nullable NSString *)stackTrace plCrashReport:(nonnull NSString *)plCrashReport completionHandler:(void (^ _Nonnull)(NSString * _Nullable message, MPExecStatus execStatus)) completionHandler;
+- (void)logBaseEvent:(nonnull MPBaseEvent *)event completionHandler:(void (^ _Nonnull)(MPBaseEvent * _Nonnull event, MPExecStatus execStatus))completionHandler;
+- (void)logEvent:(nonnull MPEvent *)event completionHandler:(void (^ _Nonnull)(MPEvent * _Nonnull event, MPExecStatus execStatus))completionHandler;
+- (void)logCommerceEvent:(nonnull MPCommerceEvent *)commerceEvent completionHandler:(void (^ _Nonnull)(MPCommerceEvent * _Nonnull commerceEvent, MPExecStatus execStatus))completionHandler;
+- (void)logNetworkPerformanceMeasurement:(nonnull MPNetworkPerformance *)networkPerformance completionHandler:(void (^ _Nullable)(MPNetworkPerformance * _Nonnull networkPerformance, MPExecStatus execStatus))completionHandler;
```

Comment:
Please @einsteinx2 correct me if Im wrong, but I see that we are passing callbacks. Are we planning to leave the callback implementation for backwards compatibility and go to an async implementation (like we are on other platforms with suspend functions, promises, futures or not?) - just a question

Reply:
Again, no method signatures are being changed in this initial work, only the object graph (mediator/eventLogging/etc)

To directly answer the question though, on iOS until we can target iOS 13 as a minimum we can't use async/await at all, so we have to use callbacks, not just as a fallback. Later when we can increase the minimum target, we'll be able to switch to async/await and either remove the callbacks entirely or more likely deprecate them and leave them until the next major version or keep them as a fallback if necessary for Obj-C users.